### PR TITLE
Make `.name` mandatory in Placeholder and SizeParam

### PR DESCRIPTION
**Pull request by inducer**
cc @matthiasdiener

### CODE REVIEW

File: pytato/codegen.py [hunk 1]
```diff
@@ -605,11 +605,7 @@ def __init__(self) -> None:
         super().__init__()
 
     def post_visit(self, expr: Any) -> None:
-        if isinstance(expr, InputArgumentBase):
-            if expr.name is None:
-                # Name to be automatically assigned
-                return
-
+        if isinstance(expr, (Placeholder, SizeParam)):
```
Comment (kaushikcfd) [inline pytato/codegen.py:608]: What if the user specifies 2 non-identical data wrappers with the same name? I think that's still a good reason to raise.

File: pytato/target/loopy/codegen.py [hunk 2]
```diff
@@ -898,7 +898,7 @@ def generate_loopy(result: Union[Array, DictOfNamedArrays, Dict[str, Array]],
     state.var_name_gen.add_names({input_expr.name
             for name in compute_order
             for input_expr in ing(outputs[name].expr)
-            if input_expr.name is not None})
+            if isinstance(input_expr, (Placeholder, SizeParam))})
```
Comment (kaushikcfd) [inline pytato/target/loopy/codegen.py:901]: User specified data wrapper names are still something we don't want to collide with.

**Review by kaushikcfd — APPROVED** [review]
Thanks, the logic of an :attr:`Placeholder.name` having the type `Optional[str]` was flaky. It's great to not hit those.

File: pytato/codegen.py (hunk 1 above)
Comment (inducer) [reply pytato/codegen.py]: cde46d1185dddc8e748dcc26d74dca5c923bd514

File: pytato/target/loopy/codegen.py (hunk 2 above)
Comment (inducer) [reply pytato/target/loopy/codegen.py]: cde46d1185dddc8e748dcc26d74dca5c923bd514

**Pull request merged**